python/cudf/cudf/core/index.py

```diff
@@ -2673,6 +2676,153 @@ def categories(self):
+    both or neither.
+
+    Returns
+    IntervalIndex
+    """
+
+    def __new__(
+        cls,
+        start: int = None,
+        end: int = None,
+        periods: int = None,
+        freq: int = None,
+        closed: str = "right",
+        name: str = None,
+    ) -> "interval_range":
+        if freq and periods and start and end:
+            raise ValueError(
+                "Of the four parameters: start, end, periods, and "
+                "freq, exactly three must be specified"
+            )
+        elif freq and periods and start:
+            end = freq * periods + start
+            left_col = cupy.arange(start, end, freq)
+            right_col = cupy.arange(start + freq, end + 1, freq)
+            if len(left_col) != len(right_col):
+                left_col = cupy.arange(start, end - freq, freq)
+            if len(right_col) == 0 or len(left_col) == 0:
+                return pd.IntervalIndex([], closed=closed)
+        elif freq and periods and end:
+            start = end - (freq * periods)
+            left_col = cupy.arange(start, end, freq)
+            right_col = cupy.arange(start + freq, end + 1, freq)
+            if len(left_col) != len(right_col):
+                left_col = cupy.arange(start, end - freq, freq)
+            if len(right_col) == 0 or len(left_col) == 0:
+                return pd.IntervalIndex([], closed=closed)
+        elif freq and not periods:
+            left_col = cupy.arange(start, end, freq)
+            if end is not None:
+                end = end + 1
+            if start is not None:
+                new_freq = start + freq
+            elif start is None:
+                new_freq = freq
+            right_col = cupy.arange(new_freq, end, freq)
+            # sometimes the left col overlaps with the right col
+            if len(left_col) != len(right_col):
+                if end is not None:
+                    end = end - freq
+                left_col = cupy.arange(start, end, freq)
+            if len(right_col) == 0 or len(left_col) == 0:
+                # I get an error passing an empty list to cudf IntervalIndex
+                return pd.IntervalIndex([], closed=closed)
+        elif periods and not freq:
+            if end is not None:
+                end = end + 1
+            periods_array = cupy.asarray(cupy.arange(start, end))
+            hist, bin_edges = cupy.histogram(periods_array, periods)
+            # cupy.histogram turns all arrays into a float array
+            # this can cause the dtype to be a float instead of an int
+            # the below adjusts for this
+            if cupy.all(cupy.mod(bin_edges, 1) == 0):
+                bin_edges = bin_edges.astype(int)
+            left_col = bin_edges[:-1]
+            right_col = bin_edges[1:]
+
+        interval_col = column.build_interval_column(
+            left_col, right_col, closed=closed
+        )
+        return cast(interval_range, IntervalIndex(interval_col))
```

Comment:
I've been getting an error with mypy when I return `IntervalIndex` from this function. This is the error I get 

`Incompatible return type for "__new__" (returns "IntervalIndex", but must return a subtype of "interval_range")`

To get the test to pass I used cast and made it an `interval_range` type instead. Is there a better way to fix this error @shwina?

Reply:
Would it work to make `interval_range` a function that returns a new `IntervalIndex` instead of a class with a `__new__` method? I don't believe `interval_index` is a type in Pandas but I could be wrong :)

Reply:
Yes, that makes a lot of sense! I'll do that :D